feat: add Argon2 parameter blog post (#48)
ory/web: src/pages/blog/argon2-parameters-choice-best-practice.md

---
published: true
path: '/argon2-parameter-choice-best-practice/'
title: >

>   Choose Argon2 parameters following best practices

Reviewer: ```suggestion
  Set-up Argon2 for secure password hashing 
```

metaDescription: >

>   Learn how to choose Argon2 parameters following best practices

Reviewer: ```suggestion
  How to configure Argon2's cryptographic hashing function 
```

Author: Argon2 itself is the hashing function
What about
```suggestion
  How to configure the cryptographic hashing function Argon2 following best practices
```

Reviewer: LGTM.

metaTitle: >

>   Choose Argon2 parameters following best practices

Reviewer: ```suggestion
  Set-up Argon2 for secure password hashing
```

publishedAt: '2020-11-02'
author: 'Patrik Neu'
overline: >
  Security Best Practices

category: Security
subtitle: >

>   Learn how to choose Argon2 parameters following best practices

Reviewer: ```suggestion
  How to configure Argon2's cryptographic hashing algorithm
```

teaser:

>   This article sums up all the security best practices around Argon2 parameter

Reviewer: ```suggestion
  This article sums up all the security best practices around Argon2's secure password hashing algorithm, specifically in regards to parameter
```

Author: Same here, argon2 is the hash function
```suggestion
    This article sums up all the security best practices around the Argon2 secure password hashing algorithm, specifically in regards to parameter
```

Reviewer: LGTM.

  choice. It also presents a CLI tool to automatically calibrate the best
  values.
---

## What is Argon2?

To securely manage credentials, one has to only store a cryptographic hash of
the credential. Whenever the credential has to be checked, the hash of the
provided credential is computed and compared to the stored hash. This ensures
that no one is able to retrieve the credentials, even with full access to the
services' storage.

Argon2 is a cryptographic hash algorithm specifically designed to secure
passwords. It is
[recommended by OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id)
in the Argon2id variant as a modern, secure and flexible algorithm. This
flexibility means that one has to choose some parameters and is probably the
reason why you are reading this article.

> ## Meet the Argon2 Parameters

Reviewer: ```suggestion
## Meet the Argon2 cryptographic hashing Parameters
```

---
published: true
path: '/argon2-parameter-choice-best-practice/'
title: >
  Determine Argon2 parameters for secure password hashing 

metaDescription: >
  How to configure the cryptographic hashing function Argon2 following best
  practices

metaTitle: >
  Argon2 parameters for secure password hashing

publishedAt: '2020-11-04'
author: 'Patrik Neu'
overline: >
  Security Best Practices

category: Security
subtitle: >
  Learn how to choose Argon2 parameters for cryptographic password hashing
  following security best practices

teaser:
  This article sums up all the security best practices around Argon2 parameter
  choice. It also presents a CLI tool to automatically calibrate the best
  values.

> ---

Author: I think we need to add that in the template, I tracked it as https://github.com/ory/web/issues/50

## What is Argon2?

To securely manage credentials, one has to only store a cryptographic hash of
the credential. Whenever the credential has to be checked, the hash of the
provided credential is computed and compared to the stored hash. This ensures
that no one is able to retrieve the credentials, even with full access to the

> services' storage.

Reviewer 2: Consider 
even with full access to the system's' storage.
<even with full access to the services' storage.>

Argon2 is a cryptographic hash algorithm specifically designed to secure
passwords. It is
[recommended by OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id)
in the Argon2id variant as a modern, secure and flexible algorithm. This
flexibility means that one has to choose some parameters and is probably the
reason why you are reading this article.

> 

Reviewer 2: ADD
NB After yesterday's discussion with @zepatrik, this might not be in the scope...

By way of example, I first want to deconstruct a typical Argon2 hash and explain the contents.  To visualise the process, I wrote a very simple application (github.com/ory/oryworks/argon2ui.go or js) that has the following:

type credential struct {
	user string `json:"user"`
	pw string `json:"password"`
	salt string `json:"salt"`
}
type hashParams struct {
         iterations int `json:"iteration"`
         hlen string `json:"hashlength"`
         hashalgo string `json:"Argon2id"`
...
}

#funct main ()
h := patrikshashingfunction(using credential, hashParams)

return h

$argon2id$v=19$m=16,t=2,p=1$MHJ5S3JAdG9z$btlJ1wXz/hD81T6XZ9ZuPQ

hashalgo := argon2id
salt := v=19$m=16,t=2,p=1

## Meet the Argon2 Cryptographic Hashing Parameters

Let me introduce you to the parameters and explain their impact on the hashing

> operation. This section is based on the

Reviewer 2: Consider
role in the on the hashing computation.
<impact on the hashing operation.>

[Argon2 specification paper](https://password-hashing.net/argon2-specs.pdf).

1. Memory: The memory used by the algorithm. To make hash cracking more
   expensive for an attacker, you want to make this value as high as possible.
2. Iterations: The number of iterations over the memory. The execution time is

>    linearly dependent on this parameter. It allows you to increase the

Reviewer 2: Correction
The execution time is linearly correlated on the number of iterations

   computational cost required to calculate one hash.
3. Parallelism: The number of threads to use. This should be chosen as high as
   possible to reduce the threat imposed by parallelized hash cracking.
4. Salt Length: The authors of Argon2 recommend this parameter to be 128 bits,
   but say it can be reduced to 64 bits in the case of space constraints.

> 5. Key Length (i.e. Hash Length): This parameter depends on the intended usage.

Reviewer 2: Correction @zepatrik this is not clear
Key Length: This parameter defines the byte size in bits??? As such it defines the hash length in bytes and depends on the encoding method

Author: It depends on the implementation. Some might use bits, others bytes. In general this is not fixed and as you said it also depends on the encoding. I would just not mention any unit there, as there is not a fixed one.

Reviewer 3: ```suggestion
5. **Key Length** (i.e. Hash Length): This parameter depends on the intended usage.
```

>    The authors say a value of 128 bits should be sufficient for most

Reviewer 2: ADD
The Argon2 algorithm authors claim ...

   applications. If you plan to use the hash as a derived key for e.g. AES, you
   can use this parameter to get a key of the required length.

## How to Choose the Parameter Values

Now that we know the parameters, we can start to determine the exact values.
Start with the fixed parameters. The degree of **parallelism** should be twice
the amount of available CPU cores dedicated to hashing. Choose a **salt length
and key length** of 128 bits unless you have a strict space constraint or
require longer keys.

### Desired Execution Time

Our goal is to tune the parameters so that a single hashing operation takes an
acceptable amount of time. Here, user experience is in conflict with security,
and even in the interest of users security should win out. For frontend
applications the execution time should be at least 0.5s, but you should strive
to make it 1s. Backend authentication can take a bit longer, but that depends on
your use case. Have a look in the
[Argon2 specification paper](https://password-hashing.net/argon2-specs.pdf) to
get some recommended durations for different applications.

### Adjust Memory and Iterations

> To reach the desired execution time, you have two variables to tweak. It is

Reviewer 2: CONSIDER
you can tweak two variables...

recommended to start with the highest amount of memory possible and one
iteration. Reduce the memory until one hashing operation takes less than your
desired duration. Next, advance the number of iterations to approach the desired
execution time as close as possible.

## Automate it!

If the previous paragraph sounds like an algorithm to you, then you are
absolutely right. We wrote a small CLI helper that allows you to run this
procedure in an automated manner. It is part of
[ORY Kratos](https://github.com/ory/kratos). You can use the prebuild binary or
docker image to run the CLI on your server and figure out the best values for
your setup. It is as easy as running:

```
$ kratos hashers argon2 calibrate 1s
```

Or using docker:

```
$ docker run -it --entrypoint kratos oryd/kratos:latest hashers argon2 calibrate 1s
```

The CLI allows you to set all your constrains, just have a look at all the
options using the `--help` flag. An exemplary output could be:

```yaml
{
  'memory': 1048576, # = 1GB in KB
  'iterations': 2,
  'parallelism': 8,
  'salt_length': 16, # in bytes
  'key_length': 32, # in bytes
}
```

> 

Reviewer 2: ADD

## Conclusion
This article is purposely brief and aims to be an introduction and reference for developers that want to use the Argon2 password hashing algorithm.

Reviewer 3: ```suggestion
## Conclusion

This article is purposely brief and aims to be an introduction and reference for developers that want to use the Argon2 password hashing algorithm for implementing secure login. Check out the [Ory Kratos](https://github.com/ory/kratos) open source project if you are looking for a login, registration, 2fa, profile management system where you can bring your own UI!
```

## References

- [Argon2 IEFT draft, section on parameter choice](https://tools.ietf.org/html/draft-irtf-cfrg-argon2-11#section-4)
- [Argon2 Specification Paper](https://password-hashing.net/argon2-specs.pdf)
- [OWASP Password Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html)

> - [Calibration CLI Implementation in C# by @bburman](https://github.com/bburman/Twelve21.PasswordStorage)

Reviewer 3: We will be adding these sections as part of: https://github.com/ory/web/issues/5

---
published: true

> path: '/argon2-parameter-choice-best-practice/'

Reviewer 3: ```suggestion
path: '/choose-recommended-argon2-parameters-password-hashing/'
```

Reviewer 3: File name must also be updated

title: >

>   Determine Argon2 parameters for secure password hashing 

Reviewer 3: ```suggestion
  Choose Argon2 Parameters for Secure Password Hashing and Login
```

metaDescription: >

>   How to configure the cryptographic hashing function Argon2 following best
>   practices

Reviewer 3: ```suggestion
  Pick recommended Argon2 (Argon2id, Argon2i) parameters (iteration, memory, parallelism) for secure login and password hashing, following security best practices using only open source.
```

metaTitle: >

>   Argon2 parameters for secure password hashing

Reviewer 3: ```suggestion
  Recommended Argon2 Parameters for Secure Login and Protecting Passwords.
```

publishedAt: '2020-11-04'
author: 'Patrik Neu'
overline: >

>   Security Best Practices

Reviewer 3: ```suggestion
  Secure Password Hashing and Login
```

category: Security
subtitle: >

>   Learn how to choose Argon2 parameters for cryptographic password hashing
>   following security best practices

Reviewer 3: ```suggestion
  Use open source to pick recommended Argon2 (Argon2id, Argon2i) parameters (iteration, memory, parallelism) and settings for secure login and password hashing.
```

teaser:
  This article sums up all the security best practices around Argon2 parameter
  choice. It also presents a CLI tool to automatically calibrate the best
  values.
---

> ## Summary

Reviewer 3: Please avoid headlines following the hero section

```suggestion
```

At Ory we develop open source access control and user management software. This
guide sums up all the security best practices we follow and developed around
Argon2. It first provides some background on Argon2 and best practice for
choosing its parameters. I also wrote a CLI that allows you to calibrate the
Argon2 parameter values according to your constrains and resources.

> ## What is Argon2?

Reviewer 3: ```suggestion
## Arogn2 - Secure Login and Password Hashing
```

To securely manage credentials, one has to only store a cryptographic hash of
the credential. Whenever the credential has to be checked, the hash of the
provided credential is computed and compared to the stored hash. This ensures
that no one is able to retrieve the credentials, even with full access to the
system's storage.

Argon2 is a cryptographic hash algorithm specifically designed to secure
passwords. It is
[recommended by OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id)
in the Argon2id variant as a modern, secure and flexible algorithm. This
flexibility means that one has to choose some parameters and is probably the
reason why you are reading this article.

> ## Meet the Argon2 Cryptographic Hashing Parameters

Reviewer 3: ```suggestion
## Argon2's Cryptographic Password Hashing Parameters
```

Let me introduce you to the parameters and explain their role in the hashing
operation. This section is based on the

> [Argon2 specification paper](https://password-hashing.net/argon2-specs.pdf).

Reviewer 3: ```suggestion
paper [Argon2: the memory-hard function for password hashing and other applications](https://password-hashing.net/argon2-specs.pdf).
```

> 1. Memory: The memory used by the algorithm. To make hash cracking more

Reviewer 3: ```suggestion
1. **Memory**: The memory used by the algorithm. To make hash cracking more
```

   expensive for an attacker, you want to make this value as high as possible.

> 2. Iterations: The number of iterations over the memory. The execution time

Reviewer 3: ```suggestion
2. **Iterations**: The number of iterations over the memory. The execution time
```

   correlates linearly with this parameter. It allows you to increase the
   computational cost required to calculate one hash.

> 3. Parallelism: The number of threads to use. This should be chosen as high as

Reviewer 3: ```suggestion
3. **Parallelism**: The number of threads to use. This should be chosen as high as
```

   possible to reduce the threat imposed by parallelized hash cracking.

> 4. Salt Length: The authors of Argon2 recommend this parameter to be 128 bits,

Reviewer 3: ```suggestion
4. **Salt Length**: The authors of Argon2 recommend this parameter to be 128 bits,
```

   but say it can be reduced to 64 bits in the case of space constraints.
5. Key Length (i.e. Hash Length): This parameter depends on the intended usage.
   The Argon2 algorithm authors claim that a value of 128 bits should be
   sufficient for most applications. If you plan to use the hash as a derived
   key for e.g. AES, you can use this parameter to get a key of the required
   length.

> ## How to Choose the Parameter Values

Reviewer 3: ```suggestion
## Choose the Right Argon2 Parameters
```

Now that we know the parameters, we can start to determine the exact values.
Start with the fixed parameters. The degree of **parallelism** should be twice
the amount of available CPU cores dedicated to hashing. Choose a **salt length
and key length** of 128 bits unless you have a strict space constraint or
require longer keys.

> ### Desired Execution Time

Reviewer 3: ```suggestion
### Login Time Versus Security
```

Author: The idea here was that the execution time is a (meta) parameter that you want to tweak. That's why it is under `## Choose the Right Argon2 Parameters`

Our goal is to tune the parameters so that a single hashing operation takes an
acceptable amount of time. Here, user experience is in conflict with security,
and even in the interest of users security should win out. For frontend
applications the execution time should be at least 0.5s, but you should strive
to make it 1s. Backend authentication can take a bit longer, but that depends on
your use case. Have a look in the
[Argon2 specification paper](https://password-hashing.net/argon2-specs.pdf) to
get some recommended durations for different applications.

> ### Adjust Memory and Iterations

Reviewer 3: ```suggestion
### Adjust Memory and Iterations Parameters
```

To reach the desired execution time, you can tweak two variables. It is
recommended to start with the highest amount of memory possible and one
iteration. Reduce the memory until one hashing operation takes less than your
desired duration. Next, advance the number of iterations to approach the
desired< execution time as close as possible.

> ## Automate it!

Reviewer 3: ```suggestion
## Use Open Source to Choose Recommended Parameters
```

If the previous paragraph sounds like an algorithm to you, then you are
absolutely right. We wrote a small CLI helper that allows you to run this
procedure in an automated manner. It is part of our user management system
[Ory Kratos](https://github.com/ory/kratos). You can use the prebuild binary
from GitHub releases or docker image to run the CLI on your server and figure
out the best values for your setup. It is as easy as running:

```
$ kratos hashers argon2 calibrate 1s
```

Or using docker:

```

> $ docker run -it --entrypoint kratos oryd/kratos:latest hashers argon2 calibrate 1s

Reviewer 3: Please fix the version here!

Author: We did not release the CLI yet

Reviewer 3: Then this won't work with the latest tag either though?

Author: Sure, I assumed we post this with the next release and fix the version here beforehand. Should have tracked it somewhere.

Reviewer 3: ok let's tag it as the next release, we can do the release today

published: true
path: '/argon2-parameter-choice-best-practice/'
title: >

>   Set-up Argon2 for secure password hashing 

Author: Parameter is definitely a keyword for this blog post
```suggestion
  Determine Argon2 parameters for secure password hashing 
```

Reviewer: LGTM

metaDescription: >
  How to configure the cryptographic hashing function Argon2 following best practices

metaTitle: >

>   Set-up Argon2 for secure password hashing

Author: Parameter is definitely a keyword for this blog post

```suggestion
  Determine Argon2 parameters for secure password hashing
```

Reviewer: LGTM

category: Security
subtitle: >

>   How to configure Argon2's cryptographic hashing algorithm

Author: Argon2 is the algorithm
```suggestion
  How to configure Argon2 parameters for cryptographic password hashing
```

Reviewer: LGTM

teaser:

>   This article sums up all the security best practices around Argon2's secure password hashing algorithm, specifically in regards to parameter

Author: Revert as this is wrong
```suggestion
  This article sums up all the security best practices around Argon2 parameter
```

the credential. Whenever the credential has to be checked, the hash of the
provided credential is computed and compared to the stored hash. This ensures
that no one is able to retrieve the credentials, even with full access to the

> services storage.

Author: This is genitive, as it is the storage of the service, please revert

Reviewer 4: Tbh this is more a personal preference as I don't really like the possessive apostrophe.

Author: Not sure about that, let's have a native speaker decide that.

Reviewer: Personally, I would remove the apostrophe.😘

flexibility means that one has to choose some parameters and is probably the
reason why you are reading this article.

> ## Meet the Argon2 cryptographic hashing Parameters

Author: I feel like these are two keywords too many, keep in mind that people still have to read this
```suggestion
## Meet the Argon2 Parameters
```

Also headings should have this weird casing, don't know the name
```suggestion
## Meet the Argon2 Cryptographic Hashing Parameters
```

Reviewer: LGTM

Author: Guys which one? You both answered an **OR** question with :+1: :facepalm:

Reviewer: 2nd one. Sorry.
